r/R/csv.R

```diff
@@ -782,12 +782,16 @@ write_csv_arrow <- function(x,
     tryCatch(
       x <- as_record_batch_reader(x),
       error = function(e) {
-        abort(
-          paste0(
-            "x must be an object of class 'data.frame', 'RecordBatch', ",
-            "'Dataset', 'Table', or 'RecordBatchReader' not '", class(x)[1], "'."
+        if (grepl("Input data frame columns must be named", conditionMessage(e))) {
+          abort(conditionMessage(e), parent = NA)
```

Comment:
I'm not sure `parent = e` does exactly what you think; here's the error with `parent = NA`:
```
Error in write_csv_arrow(x, tempfile()) : 
  Input data frame columns must be named
ℹ Column names are NULL
```

and here's the error with `parent = e`:

```
Error in write_csv_arrow(x, tempfile()) : 
  Input data frame columns must be named
ℹ Column names are NULL
Caused by error:
! Input data frame columns must be named
ℹ Column names are NULL
```

I think `parent = NA` is a better fit in all cases.  Current error in the other condition with no parent specified:

```
Error in value[[3L]](cond) : 
  x must be an object of class 'data.frame', 'RecordBatch', 'Dataset', 'Table', or 'RecordBatchReader' not 'function'.
```

With `parent = NA`:
```
Error in write_csv_arrow(c, tempfile()) : 
  x must be an object of class 'data.frame', 'RecordBatch', 'Dataset', 'Table', or 'RecordBatchReader' not 'function'.
```

With `parent = e`:
```
Error in write_csv_arrow(c, tempfile()) : 
  x must be an object of class 'data.frame', 'RecordBatch', 'Dataset', 'Table', or 'RecordBatchReader' not 'function'.
Caused by error in `inherits()`:
! argument "schema" is missing, with no default
```

I've updated it both so they have `parent = NA`.

Reply:
> the problem is that the original error is not the one you expected 

I don't understand; the original error is exactly the one I expected, and I know this because I've used `grepl()` to test its contents.

Reply:
I can see your objections to the way it's being handled there, but that code was already before, and the scope of this PR was just to add in the improved error messaging in specific circumstances, which has been achieved.  Good call opening a new ticket, I think that makes the most sense here.